lib/kubernetes-deploy/kubernetes_resource/redis.rb

```diff
@@ -0,0 +1,80 @@
+module KubernetesDeploy
+  class Redis < KubernetesResource
+    TIMEOUT = 5.minutes
+    UUID_ANNOTATION = "redis.stable.shopify.io/owner_uid".freeze
+
+    def initialize(name, namespace, context, file)
+      @name, @namespace, @context, @file = name, namespace, context, file
+    end
+
+    def sync
+      _, st = run_kubectl("get", type, @name)
+      @found = st.success?
+      @status = if redis_deployment_exists? && redis_service_exists?
+        "Provisioned"
+      else
+        "Unknown"
+      end
+
+      log_status
+    end
+
+    def deploy_succeeded?
+      redis_deployment_exists? && redis_service_exists?
+    end
+
+    def deploy_failed?
+      false
+    end
+
+    def tpr?
+      true
+    end
+
+    def exists?
+      @found
+    end
+
+    private
+    def redis_deployment_exists?
+      deployment, st = run_kubectl("get", "deployments", "redis-#{redis_resource_uuid}", "-o=json")
+
+      if st.success?
+        parsed = JSON.parse(deployment)
+
+        if parsed.fetch("status", {}).fetch("availableReplicas", -1) == parsed.fetch("status", {}).fetch("replicas", 0)
+          # all redis pods are running
+          return true
+        end
+      end
+
+      false
+    end
+
+    def redis_service_exists?
+      service, st = run_kubectl("get", "services", "redis-#{redis_resource_uuid}", "-o=json")
+
+      if st.success?
+        parsed = JSON.parse(service)
+
+        if parsed.fetch("spec", {}).fetch("clusterIP", "") != ""
+          return true
+        end
+      end
+
+      false
+    end
+
+    def redis_resource_uuid
+      return @redis_resource_uuid if defined?(@redis_resource_uuid) && @redis_resource_uuid
```

Comment:
Wouldn't the `@redis_resource_uuid ||= begin` version of memoization be more straightforward, since you do want to re-run the code below when it is defined but nil?